### ISSUE TITLE
WH Loss Term

### ISSUE THREAD

**Issue by stevethatsmyname** [experiencor/keras-yolo2#256]
Unless I am reading the code incorrectly, it looks like the box WH loss term is defined as:

sum[ ( truth_wh - pred_wh) ^2 ]

whereas the yolo_v1 paper defines the loss as

sum [  (sqrt(truth_wh) -  sqrt(pred_wh) ) ^2]
[Note added sqrt]

Am I missing something in the code or is this an oversight? As mentioned in the paper this helps to prevent large boxes from dominating the loss over small boxes.

**Comment by stevethatsmyname** [comment]
Thanks for the reference. 

I just dug through the git repository (Hadn't really looked at any of the source code up until now!) and it looks like this loss term has changed quite a bit since YoloV1, but hasn't been mentioned in any of the papers.

As far as I can tell, it looks like the current incarnation is effectively using  something like:
(Log(truth) - Log(pred) )^2

If you go into yolo_layer.c line 93-107 there is a function called delta_yolo_box.

The tw and th get passed into a log.  And they take the raw pred_w and pred_h (before they have gone through an exponential).  Kinda weird and not super readable but I think that's whats happening here.  I haven't compiled this code or looked at it in debugger so it's hard to really be sure that's what's happening.

**Comment by experiencor** [comment]
@stevethatsmyname Yes, you are right about this. I should have taken the log before computing the loss. I do get it right in keras-yolo3.

**Comment by stevethatsmyname** [comment]
Cool, good to know I'm not crazy 😀 . I did a test where I trained on my dataset where i compared the end accuracy (mAP) obtained with all 3 cost terms (one with sqrt, one without sqrt, and one with log) and the log one did slightly better but only by about 1% mAP.
